### PR TITLE
Listener improvements (Fixes #128)

### LINKED ISSUE #128: Listener Improvements
The listener is still quite difficult to trip in many cases (specifically on the unit). And does not feature enough feedback to ensure that the user knows what is going on at each stage.

I propose a redesign to ensure a better user experience. Mycroft HQ will put together a library of  1000 recordings by June 16 to run tests against. We should begin making adjustments to the listener (both UX and technically) in order to make for a better experience.

### PR DESCRIPTION
**Pull request by Wolfgange3311999**
Currently only a responsive recognizer is implemented where Mycroft can only be at the beginning of the sentence. [Here's some markdown on the changes.
](https://gist.github.com/Wolfgange3311999/59bbe92928b9ff978aebe7dc668e672d)
Todo list for the future:
 - Adjust the noise parameters in mic.py to wait less for phrases to complete (must use the 1000 sample audio files to decide what works best)
 - Introduce some metrics
 - Add in the ability to place Mycroft at the end of the sentence (with a setting to change that)

### CODE REVIEW

**Comment by aatchison** [comment]
This is a big PR, so let's get as many eyes on this as possible.

**Comment by jdorleans** [comment]
@Wolfgange3311999 there are some breaking changes... please, check on them.

**Comment by Wolfgange3311999** [comment]
@jdorleans @clusterfudge Ok, the checks passed.

**Comment by jdorleans** [comment]
@Wolfgange3311999 your changes are really big. Could you develop them in more details, in your initial comment? thx

**Comment by Wolfgange3311999** [comment]
@jdorleans I've written some [markdown](https://gist.github.com/Wolfgange3311999/59bbe92928b9ff978aebe7dc668e672d) to help explain all of the changes. It's written in more of a big-picture sort of way so let me know any specific parts you would like explained in more detail.

**Comment by aatchison** [comment]
Thanks!

**Comment by the7erm** [comment]
@Wolfgange3311999 I like what's outlined in the changes.  Great idea monitoring the level and only recording when there is sound.  This is awesome.  You must have really worked hard on this.

**Comment by the7erm** [comment]
@Wolfgange3311999 I'm not sure what's going on.  I cloned your fork, and that branch, checked it out, then ran the usual stuff to start mycroft service, voice and skills.

I think you might need to rebase to the upstream master.  Something has changed since you created this branch.  All I'm getting is "Sorry I didn't catch that."

```
2016-06-17 23:02:29,931 - requests.packages.urllib3.connectionpool - INFO - Starting new HTTPS connection (1): cerberus.mycroft.ai
2016-06-17 23:02:30,694 - requests.packages.urllib3.connectionpool - DEBUG - "POST /stt/google_v2?language=%3Cmycroft.metrics.MetricsAggregator%20object%20at%200x7fe38e1ebad0%3E&version=dev-5df1e5e HTTP/1.1" 500 93
2016-06-17 23:02:30,696 - mycroft.client.speech.listener - WARNING - Speech Recognition could not understand audio
2016-06-17 23:02:30,696 - SpeechClient - INFO - Speak: Sorry, I didn't catch that.
```

**Comment by allank** [comment]
I get the same as @the7erm 

```
2016-06-18 09:26:54,621 - mycroft.client.speech.mic - DEBUG - Waiting for wake word...
2016-06-18 09:26:54,839 - requests.packages.urllib3.connectionpool - INFO - Starting new HTTPS connection (1): cerberus.mycroft.ai
2016-06-18 09:26:57,943 - requests.packages.urllib3.connectionpool - DEBUG - "POST /stt/google_v2?language=%3Cmycroft.metrics.MetricsAggregator%20object%20at%200x75c19510%3E&version=dev-5df1e5e HTTP/1.1" 500 93
2016-06-18 09:26:58,005 - mycroft.client.speech.listener - WARNING - Speech Recognition could not understand audio
2016-06-18 09:26:58,007 - SpeechClient - INFO - Speak: Sorry, I didn't catch that.
```

It also records for quite a while after I've stopped speaking, although it might be picking up ambient noise in the background?

```
2016-06-18 09:26:05,979 - mycroft.client.speech.mic - DEBUG - Recording...
2016-06-18 09:26:05,979 - SpeechClient - INFO - Begin Recording...
2016-06-18 09:26:33,176 - SpeechClient - INFO - End Recording...
```

**Comment by Wolfgange3311999** [comment]
@the7erm @allank OK, it should be fixed now. Can you verify that it is? Thanks.

**Comment by ryanleesipes** [comment]
This is working properly now, and LGTM. Merging.

**Comment by the7erm** [comment]
@Wolfgange3311999 I think something about this does a DDOS to the stt server. 

This does an excellent job of detecting the wake word.  It's the time after that I experience challenges.  There's no "go ahead" which is ok, but at the same time it seems like the response is off.  

The first few commands work great.  After about 3 or so something goes really off.  It's like it takes 30 seconds before anything even gets posted.  It could just be that the stt server is extra slow today.  I'm just not sure.

**Comment by aatchison** [comment]
I'm wondering the same thing.

File: mycroft/client/speech/main.py [hunk 1]
```diff
@@ -35,14 +35,14 @@
 config = ConfigurationManager.get()
 
 
-def handle_listening():
-    logger.info("Listening...")
-    client.emit(Message('recognizer_loop:listening'))
+def handle_record_begin():
+    logger.info("Begin Recording...")
+    client.emit(Message('recognizer_loop:record_begin'))
```
Comment (seanfitzgeraldsc) [inline mycroft/client/speech/main.py]: By emitting this event in the handler, of this event, you are creating a message storm that will eventually bring the recognizer loop to a halt. You do the same thing the record_end event handler. This is the cause of the DDOS that @the7erm is talking about.

**Comment by seanfitzgeraldsc** [comment]
This change is not ready for master. I am requesting a revert so that master is in a functional state.

File: mycroft/client/speech/main.py (hunk 1 above)
Comment (clusterfudge) [inline mycroft/client/speech/main.py]: By emitting this event in the handler, of this event, you are creating a message storm that will eventually bring the recognizer loop to a halt. You do the same thing the record_end event handler. This is the cause of the DDOS that @the7erm is talking about.

**Comment by clusterfudge** [comment]
Sorry, accidentally posted these comment originally under my work gh account.

This change break master. I am requesting an immediate revert until it can be more thoroughly reviewed and tested.

**Comment by ryanleesipes** [comment]
Revert merged, please review.

**Comment by Wolfgange3311999** [comment]
@clusterfudge Let me preface this by mentioning I'm a bit confused on the functional difference between `emit('content')` and `emit(Message('content'))`. However, my understanding was that since `handle_record_begin` is set as the callback on the loop (since loop inherits from pyee.EventEmitter), it will not get triggered again when broadcasted on the client (`WebSocket`).

Also, it should be noted that the handlers are in the exact same format [as in the old listener](https://github.com/MycroftAI/mycroft-core/blob/master/mycroft/client/speech/main.py#L38).

**Comment by seanfitzgeraldsc** [comment]
@Wolfgange3311999 you're right, and I'm really sorry for my mistake. This is the exact same behavior; I read too much into @the7erm 's report and thought I saw an immediate cause. Having now stepped through in a debugger, I can confirm this is not creating a broadcast storm.

The two emit methods are separate (and poorly named) methods, where one is on the event_loop (which is a EventEmitter), and the other is the websocket client to the message bus. The main purpose of this entry point is to run the recognizer loop, and relay a subset of messages to the message bus service. If the method naming is so bad that it confuses me (the original author), then it probably needs some TLC.

Sorry for jumping the gun, and sorry I wasn't around more for the original review of this code. Someone can "revert the revert" up in here.

Please note I still have concerns about @the7erm 's reports, and we're going to work on our review process to incorporate long running tests to detect regressions like he's reporting.

**Comment by seanfitzgeraldsc** [comment]
I've confirmed that @the7erm 's problems appear to be cerberus related (intermittent arbitrarily long/timing out request), and are not limited to TTS (I've been verifying using the user handler. Because of that, I'm unable to test this change thoroughly (short of disabling cerberus, which I might do)

**Comment by the7erm** [comment]
I've been running this voice service on this all day.  It's **VERY** responsive to the wake word now.  I'm **VERY** impressed.  :+1: and ultra high kudos go to @Wolfgange3311999 !

**Comment by ryanleesipes** [comment]
It's a big change from the old way, I've actually written a blog post on why this is a big deal. Thanks for the feedback Erm, and great job Wolf. We put a lot of thought into this change.

**Comment by Wolfgange3311999** [comment]
@clusterfudge I see it as much better to do a revert and then revert of a revert than to risk merging something with a potential bug that large :). I should definitely follow Aatchison's advice to prefer tons of smaller PRs rather than a huge PR be like this. That way, hopefully there will be more of an opportunity for others to chime in.

Do you think `relay_to_client()` would be a good rename for the handlers? (And I am assuming that there's a way to retrieve either the message name or pass that as a parameter to the callback so that ibdividual functions don't have to be made)

**Comment by clusterfudge** [comment]
I'll have to put some thought into a refactor of those; I think emit makes sense in some cases, but isn't really clear what's happening here. 

@ryanleesipes where's this blog post?

**Pull request merged**